explainer/effect-generic-trait-declarations.md

```diff
@@ -0,0 +1,562 @@
+
+We're currently in the process of adding support for the `const` and `async`
+language features to Rust. But we're looking at various other extensions as
+well, such as generator functions, fallible functions, linearity, and more.
+These are really big extensions to the language, whose implementation will take
+on the order of years. If we want to successfully introduce these features,
+they'll need to be integrated with every other part of the language. As well as
+having wide support in the stdlib.
+
+Effect Generic Trait Declarations are a minimal language feature which enable
+traits to add support for new effects, without needing to duplicate the trait
+itself. So rather than having a trait `Into`, `TryInto`, `AsyncInto`, and the
+inevitable `TryAsyncInto` - we would declare a single trait `Into` once, which
+has support for any combination of `async` and `try` effects. This is
+backwards-compatible by design, and should be able to support any number of
+effect extensions we come up with in the future. Ensuring the language can keep
+evolving to our needs.
+
+## Guaranteeing API consistency
+
+Evolving a programming language and stdlib is pretty difficult. We have to pay
+close attention to details. And in Rust specifically: once we make a mistake
+it's pretty hard to roll back. And we've made mistakes with effects in the past,
+which we now have to work with.
+
+In Rust 1.34 we stabilized a new trait: `TryInto`. This was supposed to be the
+fallible version of the `Into` trait, containing a new associated type `Error`.
+However since Rust 1.0 we've also had the
+[`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html) trait, which
+*also* provides a fallible conversion, but has an associated type `Err`. This
+means that when writing a fallible conversion trait, it's unclear whether the
+associated type should be called `Err` or `Error`.
+
+This might seem minor, but without automation these subtle
+similar-but-not-quite-the-same kinds of differences stand out. The only way to
+ensure that different APIs in different contexts work consistently is via
+automation. And the best automation we have for this is the type system.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Trait definitions
+
+The base of Effect Generic Trait Declarations is the ability to declare traits
+as being generic over effects. This RFC currently only considers the `async`
+effect, but should be applicable to most other effects (modulo `unsafe` and
+`const`, more on that later). The way a trait is defined is by adding a
+`#[maybe(effect)]` notation.  This signals that a trait may be implemented as
+carrying the effect.  For example, a version of the `Into` trait which may or
+may not be `async` would be defined as:
+
+```rust
+#[maybe(async)]
+pub trait Read {
+    #[maybe(async)]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+}
+```
```

Comment:
The trait you talk about is `Into`, but the snipped uses `Read` (cont. from: https://mas.to/@teivel/111767978991052666)